pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -308,12 +308,25 @@ func isInstantQuery(path string) bool {
 	return strings.HasSuffix(path, instantQueryPathSuffix)
 }
 
-func defaultInstantQueryParamsRoundTripper(next http.RoundTripper, now func() time.Time) http.RoundTripper {
+func defaultInstantQueryParamsRoundTripper(next http.RoundTripper) http.RoundTripper {
 	return RoundTripFunc(func(r *http.Request) (*http.Response, error) {
-		if isInstantQuery(r.URL.Path) && !r.URL.Query().Has("time") {
+		// Check r.Form first as that's just looking up in a map, versus r.URL.Query() which parses the RawQuery again.
+		// Most of the clients send the POST urlencoded form anyway.
+		// ParseSeekerBodyForm will call the standard http.Request.ParseForm which will respect the precedence of the
+		// post form over the url query.
+		if isInstantQuery(r.URL.Path) && !r.Form.Has("time") && !r.URL.Query().Has("time") {
```

Comment:
`r.Form` is only available after ParseForm was called, and in that case it parsed query parameters too. There is no need to check both: form and URL query.

Reply:
I agree, but `r.Form.Has()` is a lookup in a map while `r.URL.Query().Has()` is parsing the RawQuery, allocating the url.Values and looking up in that map.

The first one `r.Form.Has("time")` should be a shortcut for most of the cases as `time` is usually provided so we don't need to parse the `r.URL.Query` (which is the only condition to add the default "time" param).

If the form wasn't parsed yet, we just add the default "time" in the url query, relying on the precedence of the form.

Reply:
Ah, it covers a case when `r.ParseForm` was NOT called yet.

So if `r.ParseForm` wasn't called yet, and there is no `time` query parameter, then we add query parameter and parse form ... but if body actually contained `time` parameter already, query parameter is not used ("Request body parameters take precedence over URL query string values in r.Form." from docs). I think this is too much magic.

I think if we're checking `r.Form`, we should make sure that `r.ParseForm` was called (ie. call it in this method). If we don't find `time` parameter, we can put it into query parameters, and update `Form` as well. No need for new parsing.

Reply:
If we update `r.Form` directly, then we don't need other parts of the PR?

Reply:
> I think this is too much magic.

I don't think so. It's explained in the comment and it's tested. (And it's the logic it had always, but wasn't tested and wouldn't work if form was already parsed, so that's what I'm fixing here).

> If we update r.Form directly

I think that the request body handling is already too fragile (that's the most polite word I can use for it), and I don't want to make it more fragile by skipping steps here.

These middlewares are not _handling_ the requests, they're forwarding them to something unknown, so I'd rather stick to the standard path here as much as possible.

Reply:
> so I'd rather stick to the standard path here as much as possible.

I don't see what's standard in resetting body and reparsing the form multiple times, or checking both Form and query params (that part is esp. confusing). I also don't think adding value to the form after adding query parameter is "less standard" or "more fragile". It saves us the reparse, which would end up with same Form output. (This would be especially beneficial in case of malicious queries with lot of garbage posted parameters in the body)

Both ways solve the problem, and one is simpler to follow and understand. And we still need to reset body before going to next middleware.

Reply:
I think you convinced met, PTAL at https://github.com/grafana/mimir/pull/3599